### ISSUE TITLE
Run 2.0p: outstanding phoSim issues

### ISSUE THREAD

**Issue by TomGlanzman** [LSSTDESC/DC2-production#267]
This issue collects and discusses outstanding issues with the phoSim code that we would like addressed prior to the commencement of Run 2.0p.

Note that for DC2 Run 1.2p, the versions of phoSim used were: **3.7.9** and **3.7.14**

- [x] We have noticed that the PhoSim PSF ellipticity is about 2x too large and likely falls outside of LSST SRD requirements.  See #259 for more details.
- [x] We have seen sensor-to-sensor "gain-like" variations in the phosim e-images in some 1.2p focal plane visits. There was some discussion in the [desc-dc2-eyeballs slack channel around 5 September 2018](https://lsstc.slack.com/archives/C932BQ550/p1536159558000100) and it is easy to see in some of the phosim [DC2 1.2 Focal Planes](https://portal.nersc.gov/projecta/lsst/kadrlica/eimage/). It had also been seen [before 1.2p](https://lsstc.slack.com/archives/C932BQ550/p1536178839000100). Since these are e-images it is happening before the amplifiers. The vignetting in the phosim e-images takes up some of the dynamic range of the stretch in full focal plane visualizations, so the sensor-to-sensor level variation is likely not insignificant.
- [ ] In the context of making BF correction kernels, @wmwv  saw very large inter-pixel correlations for PhoSim 1.2p in the high intensity flats.  It is so large we suspect it is not the BF effect itself (which *was* turned on for 1.2p) but rather correlations induced by the sky background optimization parameters.   We don't actually know this however and, if it remains with BF turned off, it would be good to confirm this doesn't cause any other issues with correlated noise in the sky background etc. This effect can be seen in https://github.com/LSSTDESC/DC2-production/blob/u/wmwv/bf_kernel/Notebooks/make_brighter_fatter_kernel.ipynb at Out[43].
- [x] ln examining the 2nd point above it looks like the PhoSim version number is not being correctly written into the FITS header keywords.

For reference, the official phoSim issue tracker (bitbucket) is [here](https://bitbucket.org/phosim/phosim_release/issues?status=new&status=open)

**Comment by cwwalter** [comment]
I edited the original comments to start a list.  I added instance catalog size and PSF ellipticity to start.

**Comment by cwwalter** [comment]
I recall an issue of non-understood sensor gain variation which changes from exposure to exposure.  I think maybe @adrianpope remembers details.  Adrian can you add to list?

**Comment by jchiang87** [comment]
@sethdigel Could you add an item on the sky background variation from sensor-to-sensor in a given visit please?

**Comment by TomGlanzman** [comment]
Note: I removed Chris' item regarding testing phoSim with the new, x10 larger instanceCatalog and moved that to my to-do list:  https://github.com/LSSTDESC/DC2-phoSim-4/issues/1

**Comment by adrianpope** [comment]
I added links to some of the images posted in the desc-dc2-eyeballs channel around 5 September 2018. I can't remember if this was discussed on other channels or in github issues or telecon notes.

**Comment by cwwalter** [comment]
I added a comment about the large correlations seen in the context of @wmwv 's BF kernel construction.

**Comment by adrianpope** [comment]
As for BF, for 1.2p I ran both the high-intensity and low-exposure dome flats with the same background settings as the 1.2p sky images, which I think is something close to "quick background". I could run some high-intensity BF dome flats with the newer, less-quick background settings that I think we intend to use for 2.0p.

**Comment by cwwalter** [comment]
I think 1.2p was "default", not "quick background" after issues were found before we started 1.2.  So, someone more knowledgeable (@TomGlanzman) should comment if I am correct and if there is a plan to change that before 2.0p.

**Comment by TomGlanzman** [comment]
The terminology is a little confusing...

DC2 | background 
----|----
Run 1.1p | "quick background" (various added parameters in the command file)
Run 1.2p | "optimized background" (no added parameters in the command file)

(There is, in principle, "full background" where every background photon is raytraced individually but I have never attempted to run such a configuration.)

Many problems were discovered in Run 1.1p (see, for example, https://github.com/LSSTDESC/DC2-production/issues/140), and I've heard no rumors of changing from "optimized background" for Run 2.0p.  Is there such a rumor?

**Comment by adrianpope** [comment]
I'm sorry - I must have gotten my 1.Xp versioning mixed up. I know that for the dome flats I copied whatever was used for 1.2p, so it must be "optimized" not "quick". And it might not be feasible to run high intensity flats in "full background" mode.

**Comment by TomGlanzman** [comment]
To be explicit, here is the phoSim command file used for Run 1.2p:

```# Enable centroid file
centroidfile 1

# Disable fringing
fringing 0

# Disable dirt
contaminationmode 0

# Disable PRNU (Photo Response Non-Uniformity)
pixelerror 0

activebuffer 1000

airglowvariation 0.0

clearclouds

# Number of sources handled by a single thread
sourceperthread 100```

**Comment by katrinheitmann** [comment]
No plan to go back to quick background. Still on for optimized background.

What has not been decided is B/F on or off.

**Comment by johnrpeterson** [comment]
Tom et al.:

For the third (last) issue, I think you forgot to use "backgamma 1".  That was the one conclusion from a ~year ago, with the investigation by Pierre and myself.  This is important for non-optimized background for the v3.7 series (already default for v4.0).

For the first issue, please don't worry about this.  This has been extensively studied for years and is  Chihway Chang's publications and the SRD itself has been modified by the results of PhoSim as this is what is used to determine the requirements.

The second issue, I don't know.  It seems rare, so we would need a repeatable catalog/command files where we know two chips have different levels to debug it.  In the past, this I believe came up before because checkpointing was not done properly and missing some of the sequencing because of failures.  I don't think you are doing this anymore, right?  So that's not it, right?

**Comment by TomGlanzman** [comment]
John, just to be clear, you are recommending "backgamma 1" only for flats or for all simulations?

**Comment by johnrpeterson** [comment]
all simulations.  it won't matter as much for the normals as for bright flats, but all of them.

**Comment by sethdigel** [comment]
From some of the first (y-band) focal plane images that Alex generated from the 1.2p sensor visits, some individual sensors seemed to have quite different (several percent anyway) median levels of electrons/pixel from the rest of the sensors in the visit.  I wrote a script to parse the log files for the WFD y-band visits to see if the variations related to differences in the photons generated in the simulation. 

Thirty of the visits have all 81 sensors simulated in the inner 3x3 rafts, where vignetting is not very important.  Here is a sensor-by-sensor tally of the numbers of photons that phoSim reported generating by source class, for each sensor in one of those visits (190265).  

<img width="766" alt="y1p2_190265_tally" src="https://user-images.githubusercontent.com/6035835/45707689-7460e200-bb34-11e8-9971-6c24bbf1bb8d.png">

Moon-Rayleigh and Moon-Mie are scattered moon light.  Astro is the sum of photons from all astronomical sources.  I am not exactly sure what Cosmic Rays represents.  Also included are the medians of the phoSim e-images for each sensor, scaled by 5e9 to get on the same scale.  The sensors are plotted in order R11_S00, R11_S01...R33_S22.

The Moon-Rayleigh and Moon-Mie components have more or less discrete decreases for individual sensors.  And these are compensated by increases in Airglow and Cosmic Rays.  So the total number of generated photons does not change very much across the inner 3x3 rafts, but presumably the spectrum does.  This may be the reason why the median electrons/pixel also seems to have discrete decreases for individual sensors.

This visit (190265) seems to be one of the most affected by these sensor-to-sensor variations, but the effect seems to be present at varying levels, and for different sensors, in other visits.  So far I've looked only at y-band images (all of the 1.2p WFD y-band sensor visits).

I can't say whether the effect will cause problems in the DM analysis but I think that it might, e.g., affecting sky flats.

**Comment by sethdigel** [comment]
Here's the same kind of plot, but for visit 444696, which has the greatest variations of median electrons/pixel in the central 3x3 rafts (~11%).

<img width="769" alt="y1p2_444696_tally" src="https://user-images.githubusercontent.com/6035835/45710316-2d2a1f80-bb3b-11e8-9e79-2d39e40bd591.png">

**Comment by rmandelb** [comment]
John:

> For the first issue, please don't worry about this. This has been extensively studied for years and is 
> Chihway Chang's publications and the SRD itself has been modified by the results of PhoSim as this 
> is what is used to determine the requirements.

In the first comment in this issue, there is a link to where this came up: it is based on us looking at Run 1.2p outputs and finding what appears to be excessively large PSF ellipticities compared to the latest version of the LSST SRD.  Since this came up in our test runs, we would be remiss if we do not dig in to understand what is going on.  For example, we need to check whether the runs we are looking at are anomalous in some way, e.g., too high airmass, or whether the problem persists when using a decently large set of exposures with reasonable airmass and seeing?  Currently the results appear concerning but it will take some more digging to understand the extent of this and whether it's actually a problem or not for our plans to investigate PSF modeling using these simulations.

**Comment by johnrpeterson** [comment]
Rachel-

PhoSim is the tool used to estimate ellipticity budgets for both the numbers in the SRD and other parts of the project.

john

**Comment by johnrpeterson** [comment]
Seth,

could you see if you can find two chips from one visit where there appears to be a discrepancy and then rerun them carefully with the exact same version of phosim, and see if the discrepancy exists.  if so, then we can debug it.

john

**Comment by sethdigel** [comment]
Good question about the versions.  Now that I look I see that the runs are a mix of phSim v.3.7.9 (15921) and v3.7.14 (2104), sometimes even in the same visit.  Cori login nodes seem to be going down.  I'll see how the version of phoSim matches up with the photon counts by type when I can.

**Comment by cwwalter** [comment]
@sethdigel I think the @kadrlica looked in to this in the -dc2 channel and (based on version keyword int the header) the cases where the sensors looked different were from the same PhoSim version.  Alex can probably say more about how he checked if you want to look again.

**Comment by kadrlica** [comment]
Chris is right and I have nothing deep to add. I briefly looked into this for the outlier sensors in the phosim FP images and found that they had the same version keyword as the rest of the FP.

**Comment by johnrpeterson** [comment]
also look into the logs to see if anything was different besides the version numbers (threads, job submission method, etc.).  since it looks isolated to some scattered chips on just some of the visits, it really seems like something about the job execution was fundamentally different for just a subset of chips.

also, if you have two example chips that are distinctly different, please point to them and the corresponding catalog/command files and add to this thread.  then Glenn could rerun those offline and try to figure out if this is repeatable.

**Comment by TomGlanzman** [comment]
Note that all sensor-visits simulated with phoSim v3.7.9 ran with 8 threads, while those with phoSim v3.7.14 ran with either 54 or 90 threads.  

(Recall that the point of moving to a new version was to address the problem of long-running sensor-visits.  The new version fixed an issue with threading.  I first tried 54 threads and if that was insufficient, went to 90 threads.  Even with 90, there were sensor-visits that timed out due to very bright AGNs)

**Comment by sethdigel** [comment]
Here's an updated plot for visit 190265 now showing the phoSim version (brown profile - lower level = v3.7.9 and upper level = v3.7.14).  The correspondence with the apparently discrete changes in the numbers of various photon source types (and the median electrons/pixel) is perfect. 
<img width="579" alt="y1p2_190265_tally2" src="https://user-images.githubusercontent.com/6035835/45837049-3f36ca00-bcc3-11e8-9b85-75392edf64ab.png">
I don't have a plot for 444696 - somehow my new scan over log files missed some but here's a plot for 284051 (chosen at random among the y-band visits that included all 81 sensors of the inner 3x3 rafts)
<img width="586" alt="y1p2_284051_tally2" src="https://user-images.githubusercontent.com/6035835/45837065-4958c880-bcc3-11e8-81f1-c254d9799ffa.png">
So I guess the question is whether this is an effect from how the jobs were run (the numbers of threads that Tom mentioned) or from some change in how the sky backgrounds (and cosmic rays?) are modeled in phoSim.

**Comment by johnrpeterson** [comment]
well when you switch between v3.7.9 and v3.7.14, it is as if you are regenerating the whole universe.   especially, because the whole random number sequencing was modified to be better threaded in those versions.   so when the airglow is high, when the clouds are significant, etc. will all change when you mix versions.  so that's likely it.  usually, you wouldn't see as much of a change between versions, but the v3.7.9-v3.7.14 affected everything.

**Comment by johnrpeterson** [comment]
it could be that the different versions have randomly predicted a different amount water vapor which would be prominent in y-band especially.  there are other things too, but then that would have less to do with the number generated and more to do with the amount accepted.

**Comment by egawiser** [comment]
I'm still trying to understand what I'm seeing in Seth's helpful plots.  Is this really showing that the different sources of photons, most of which I would expect to be (nearly) independent, are offsetting between e.g., Airglow and Moon-Rayleigh, such that the total photons are nearly constant?  It seems like there has to either be a neat physical explanation for that, or it's a bug somehow.  @sethdigel do similar effects occur in bands other than y?  Could this somehow come from variation in how each sensor's QE is modeled, and the Airglow and Moon-Rayleigh have opposite spectral slopes, leading to rough cancellation of the variations?  (I note that the cancellation is less severe in the PhoSim e-image median than in Summed background photons, though I'm not sure why that's the case.)

**Comment by johnrpeterson** [comment]
eric, i think its just the different version.  where you are getting various different random number for various things (like water vapor level).  if we run in the future we just have to be careful about using the same version consistently.

**Comment by cwwalter** [comment]
Hi All,

I think it is useful to remember the observation that flagged this issue.

Go here:

https://portal.nersc.gov/projecta/lsst/kadrlica/eimage/

and start to look at the PhoSim FPs.  For example look at the first 2 (#0 and #1).  You will see some sensors have a different bias level. The variation is obvious.

The key points are:

- It is sensor by sensor.
- The pattern does not repeat from visit-to-visit

When @kadrlica checked the version keywords of all the sensors in one of the focal planes that had this issue the version was the same across all sensors in the visit.

**Comment by cwwalter** [comment]
forgot to type: 

@sethdigel are you studies consistent with what Alex found in terms of consistency of version?

**Comment by sethdigel** [comment]
> are your studies consistent with what Alex found in terms of consistency of version?

No.  I think I figured out what is going on.  Alex looked at the BRANCH keyword in the FITS header of the e-images.  It has values that look like this:
`BRANCH  = '/global/common/software/lsst/cori-haswell-intel/phosim/v3.7.9'`

Instead, in parsing the photon counts from the log files for the phoSim runs, I extracted the phoSim version number from the PHOSIM_ROOT environment variable reported there.  It has values like this:
`PHOSIM_ROOT=/global/common/software/lsst/cori-knl-intel/phosim/v3.7.9`

For the first visit (12456) at the link Chris posted, I just checked, and all 81 of the sensor visits in the inner 3x3 rafts have v3.7.9 indicated in their FITS headers.  **But** from the log files, I am finding that 11 are v3.7.14.  Since these correlate with each big jump up or down in a background component, I suspect that the FITS header keyword is being set incorrectly.

Here's a plot for that visit (12456).  It is a bit more interesting case than others I have looked at.  As before the big jumps down in Rayleigh scattered photons correspond to the big jumps up in Airglow (and correspondingly large fractional increases in Cosmic Ray photons), and each of these correspond to 'dropout' images.  Two differences I see here:  sometimes the Rayleigh photons jump up a bit and the Airglow jumps down, in v3.7.9 visits.  These also seem to correspond to (smaller) changes in median electrons/pixel.  Also, the overall total number of photons (the curve at the top) does not change very much at all, much less even than for visit 190265, which I posted earlier in this thread.  Again, I am plotting just the inner 81 sensors, which span about 1 square foot of the focal plane and are looking at about 1.5x1.5 deg of sky.

<img width="595" alt="y1p2_12456_tally3" src="https://user-images.githubusercontent.com/6035835/45897546-4aa1f800-bd8c-11e8-9510-d563a3e986c1.png">

Jim reminded me yesterday that we have this mix of v3.7.9 and v3.7.14 because a number of sensor visits timed out in the initial v3.7.9 runs.  Tom reminded us that v3.7.14 worked much more efficiently with threads, and it was used to fill in these sensor visits.  I do not think the sky model code changed, and John has not indicated so.  

I have not looked at other bands.  I suspect that the issues might be less obvious at shorter wavelengths, for which the observations tend to be taken when the Moon is lower and the sky is darker.  My hypothesis is that the shifting relative contributions of the background components for the v3.7.9 and v3.7.14 runs is changing the spectrum enough in y-band to make the observed changes in the median electrons per pixel in the e-images.

I don't want to drag this on.  I think we understand that the dropouts are correlated with the phoSim version change.  If changes in background photon spectra over the focal plane don't cause insurmountable problems in DM processing, then I don't think we need to care.  I think that we'd probably be ok if we stayed with a single version of phoSim for an entire run.  I can appreciate that there are complications with threading and random number generator seeding that could make the sky backgrounds change.  I'd be more reassured if the backgrounds could be specified entirely in the configuration file, but I do not think that is in the cards, and probably would have other complications.

**Comment by johnrpeterson** [comment]
just to add:  that if the version did change (like Seth is indicated), then it is 100% clear to me that the sky background would change.  this is because we changed all the random number generation between 3.7.9 and 3.7.14 to make the threading more efficient.  this would then have the consequence of changing all sorts of time-dependent environmental things (like the water vapor, airglow intensity, etc.) and would affect the background in various ways.  in a way, this was the worst case because we changed the random number sequencing, normally differences between versions might not be as noticeable.  ideally you'd always run a data challenge with just one version, and if you ever did change it for some important practical reason then at least the visits should be a single version but visit to visit could be different.

**Comment by egawiser** [comment]
@johnrpeterson that all makes sense, but can you explain why we're seeing *offsetting* changes in different components of the background?   If it's just a change in random number generation, shouldn't we see the biggest variation between versions in the summed background photons?

**Comment by rmjarvis** [comment]
@egawiser It is because PhoSim is trying to match a target sky level given by OpSim (in r maybe?) using a model with several components.  Those components can vary, but the overall total is constrained.  In Y band, there will be some variation from the different models, but it is still fairly constrained.

**Comment by sethdigel** [comment]
By popular demand, here is a plot for i-band visit 227890, which also was mixed phoSim v3.7.9 and 3.7.14.  For this visit, the Moon does not contribute anything.  The Airglow and Cosmic Rays change a lot (and so does the overall total number of photons that phoSim generated), but without obvious impact on the median electrons/pixel.  I'd guess that they are changing at wavelengths outside the i band.
<img width="827" alt="i1p2_227890_tally3" src="https://user-images.githubusercontent.com/6035835/45903904-1d5f4500-bda0-11e8-9959-82721846ab19.png">

**Comment by egawiser** [comment]
Thanks, John!  What I was missing but which Mike kindly explained just above is that apparently, after all of the components vary from random number draws, we are later constraining the background total to a target OpSim sky level.  That's what creates the appearance of offsetting fluctuations, which was probably already understood by everyone else working on this issue.

**Comment by johnrpeterson** [comment]
well, no, it doesn't really constraint the total, so that's not it.  you can just constrain the variable part (airglow).

**Comment by adrianpope** [comment]
I'm still not good at tracking github issues, so I got behind on this, but I have a couple comments/questions/ideas:

1) Chip-to-chip gain-like variation: It sounds like we highly suspect PhoSim versioning is the cause. Would it be worth picking one (or few) focal plane visit that looked bad by eye in 1.2p and re-simulating it from scratch with a single version/configuration of PhoSim? Each sensor-visit is ~1 KNL-hour, so for 189 sensor-visits and a NERSC charge factor of ~100 it's ~10-20k NERSC-hours per focal plane visit, so not terribly expensive.

2) B/F: Should I re-run some high intensity PhoSim flat pairs with backgamma=1 (but otherwise the same 1.2p setting)? Would someone have time to analyze the outputs?

**Comment by johnrpeterson** [comment]
Adrian-

both of these were already done.

Tom addressed the first one, and yes it was due to using incompatible versions.

And Pierre and I studied the second one (see SAWG emails), and this run should have been done with backgamma=1.

john

On Nov 10, 2018, at 5:48 PM, Adrian Pope <notifications@github.com<mailto:notifications@github.com>> wrote:


I'm still not good at tracking github issues, so I got behind on this, but I have a couple comments/questions/ideas:

  1.  Chip-to-chip gain-like variation: It sounds like we highly suspect PhoSim versioning is the cause. Would it be worth picking one (or few) focal plane visit that looked bad by eye in 1.2p and re-simulating it from scratch with a single version/configuration of PhoSim? Each sensor-visit is ~1 KNL-hour, so for 189 sensor-visits and a NERSC charge factor of ~100 it's ~10-20k NERSC-hours per focal plane visit, so not terribly expensive.

  2.  B/F: Should I re-run some high intensity PhoSim flat pairs with backgamma=1 (but otherwise the same 1.2p setting)? Would someone have time to analyze the outputs?

**Comment by salmanhabib** [comment]
I would like to see explicit documentation of any statements about validation tests presented here explicitly. We have had too much confusion already. 

Also the ellipticity question needs to be addressed here with actual data from the appropriate version of PhoSim.

**Comment by salmanhabib** [comment]
To add to the above: I think Adrian should do the tests.

**Comment by rmandelb** [comment]
@salmanhabib - to assist with your request (regarding PSF ellipticity) I have tried to concisely summarize the findings from Run 1.2p that were of concern.  The summary is in #259 at the end, [this comment](https://github.com/LSSTDESC/DC2-production/issues/259#issuecomment-437952159) -- it gives numbers for what Javi et al were finding and why they were considered to be of concern, which could be the good starting point for a quantitative investigation into this question.

**Comment by jchiang87** [comment]
Just a comment on the mismatch between the `BRANCH` header keyword and the `PHOSIM_ROOT` env var listed in the log files:

The fact that the `BRANCH` keyword points at the haswell installation while the `PHOSIM_ROOT` entry in all of the log files points at the knl installation is an indication of what's going on.  In Tom's phosim workflow (@TomGlanzman can confirm), I believe that he's using the haswell nodes to set up the individual sensor-specific raytrace runs that are later executed on KNL.   He is essentially running phosim.py on haswell, but arranging for raytrace to run on knl.   This means that the phosim.py script is setting the `BRANCH` keyword in the par files that get streamed into the executables, the raytrace program in particular, and that's how raytrace is setting that keyword.  When Tom reran raytrace using v3.7.14, he apparently did not rerun the setup part, so the old `BRANCH` value pointing at haswell/v3.7.9 was carried over to the new eimage files.

**Comment by jchiang87** [comment]
Tom posted the [following](https://lsstc.slack.com/archives/C978LTJGN/p1542394401255500) in the #desc-dm-dc2 channel on Slack:

> Just an FYI:  but a new phosim test run has recently been completed: visit (obsHistID) 190265, one which Seth observed sensor-to-sensor gain variation (https://github.com/LSSTDESC/DC2-production/issues/267#issuecomment-422498827).  This test uses the same version of phoSim (3.7.16fat) for each sensor which may have addressed this issue based on Seth's quick first look.
The 'lsst_e*' and 'lsst_a*' files are located here:  /global/projecta/projectdirs/lsst/production/DC2/DC2-R2-0p-test-r/output/0000009

Here's the focalplane mosaic of the eimages for that run:
![v190265-fy-focalplane-summary](https://user-images.githubusercontent.com/1994473/48736174-66484400-ebff-11e8-9653-8cf902696b42.png)

lgtm.

**Comment by adrianpope** [comment]
Re: chip-to-chip gain-like variations...

For reference, here's what that eimage focalplane mosaic looked like in 1.2p:
![v190265-fy-focalplane-summary](https://user-images.githubusercontent.com/3528672/48871282-e2c25a80-ed98-11e8-81f6-f6c4043e7c85.png)

For posterity, here's a [link to that eimage focalplane mosaic](https://portal.nersc.gov/projecta/lsst/kadrlica/eimage/PhoSim/v190265-fy-focalplane-summary.png) in the [1.2 focal plane checker](https://portal.nersc.gov/projecta/lsst/kadrlica/eimage/).

In slack, Seth also posted the [histogram](https://lsstc.slack.com/archives/C978LTJGN/p1542398213278400) of median pixel values in 1.2p:
<img width="566" alt="y190265_meds" src="https://user-images.githubusercontent.com/3528672/48871293-eeae1c80-ed98-11e8-846f-2913e6dfe1bf.png">

And an equivalent [histogram](https://lsstc.slack.com/archives/C978LTJGN/p1542398421279100) for the new focalplane:
<img width="626" alt="medians_inner3x3_190265" src="https://user-images.githubusercontent.com/3528672/48871389-667c4700-ed99-11e8-8fb0-75f3ba3db541.png">

I think these are the median pixel values in the eimages and @sethdigel should correct me if that's wrong.

So, passing this test is another encouraging indication that differing PhoSim versions are the cause of this effect. I don't know of any plans for further tests of this issue.

**Comment by rmjarvis** [comment]
>  I don't know of any plans for further tests of this issue.

I believe that was the sufficient test.  Thanks.

**Comment by katrinheitmann** [comment]
Thanks all! That's 3 out of 4... What should we do about the last one?

**Comment by jchiang87** [comment]
It sounds like we would want to generate high intensity flats with B/F turned off, but with the sky background parameters at their proposed settings (I think this was already suggested at Monday's CI meeting).  If the B/F kernel derived from those data still has pixel-to-pixel correlations, then we know there is a problem, and we can at least try to quantify it.

**Comment by adrianpope** [comment]
Right - I'm gearing up to re-run the high intensity PhoSim flats, plan to post some details for verification soon.

**Comment by adrianpope** [comment]
Re: B/F and background pixel correlations...

Based on my previous experience with shallow and deep PhoSim flats, I'm starting with instance catalogs that look like this for a ~70k e-/pixel exposure in i-band:

```
mjd 59580
telconfig 2
rottelpos 0.0
minsource 0
nsnap 1
domewav 0.0
domeint 18.0
obshistid 5370000
seed 5370000
filter 3
vistime 196.78
```
And a physics override command file based on 1.2p but with no cosmic rays, no telescope jitter, and backgamma=1:
```
# commands.txt - phoSim commands/physics-overrides

# Enable centroid file
centroidfile 1

# Disable fringing
fringing 0

# Disable dirt
contaminationmode 0

# Disable PRNU (Photo Response Non-Uniformity)
pixelerror 0

activebuffer 1000

airglowvariation 0.0

clearclouds

# Number of sources handled by a single thread
sourceperthread 100

# turn off cosmic rays?
raydensity 0.0

# turn off telescope position jitter
rotationjitter 0.0
azimuthjitter 0.0
elevationjitter 0.0

backgamma 1
```
Questions:

1. Tree Rings: I think we want to turn off fixed detector patterns for these tests. What are the commands for this?

2. Brighter/Fatter: Should we start with the B/F effect turned off in order to study the pixel correlations of the backgamma=1 background? What are the commands for this? 

3. Sensor Effects: Some PhoSim timing tests for 2.1p were run with “chargesharing 0” and “fieldanisotropy 0”. What do those do, and how should they be set for these tests?

4. Dark Current: I’ve been keeping domeint=18.0 and adjusting exposure time to get the desired e-/pixel, but that means exposure times of ~100s of seconds for ~70-90k e-/pixel. When should I worry about dark current, eg. would it be better to use a constant short exposure time and alter domeint to set e-/pixel? Does dark current show up in the e-image or just the a-images?

5. Originally I ran pairs of flats for {70k, 80k, 90k} e-/pixel in {u, i} bands for the R22_S11 sensor. Should I re-run the same things, or start with a sub-set? I wasn’t sure if we wanted all of this without B/F, and then some/all with B/F turned on.

6. Who can analyze the outputs? I think @wmwv did this previously.

And I plan to add scripts and input files for B/F flats to the [DomeFlats_1.2p](https://github.com/LSSTDESC/DomeFlats_1.2p) repo.

**Comment by jchiang87** [comment]
I don't think tree rings and brighter/fatter can be easily toggled independently.  My understanding is that chargesharing 0 turns them both off.  Regarding fieldanisotropy, you'd need to ask John, but if it is on for the sky images, I'd leave it on for these tests.  Dark current should just be present in the amplifier images.  Unless the dark current is set at something unrealistic, it should be negligible in 100s dome flats.  I think the flat pair signal levels you ran for u and i before should be fine, but @wmwv may have a suggestion for something better for these tests.

**Comment by adrianpope** [comment]
Re: darkcurrent, I've since found that there is some dark current info in the FITS headers of the outputs, so I can see what that was for the previous deep phosim dome flats. I remember dark current being an issue on some early ImSim deep flats when the exposure times were incredibly long, but I don't remember what exposure time the ImSim deep flats settled on, and I was thinking I might try to make it similar with phosim since the B/F analysis went well eventually with ImSim.

Re: fieldanisotropy, it was on for the sky images, but the only examples I have to look at toggle both fieldanisotropy and chargesharing at the same time (for some non-production tests).

So is the proposal that I start with (1.2p + backgamma=1 + chargesharing=0) and run all the {u,i} x {70k,80k,90k} pairs with those settings? Then depending on what that looks like we might try chargesharing=1 which will re-enable tree rings and B/F?

**Comment by cwwalter** [comment]
As I recall chargesharing controls BF.  In this thread where we went through this before:

https://github.com/LSSTDESC/SSim_DC1/issues/27

I asked John if cleardefects also clears treerings and he said 'yes' and that "cleardefects includes chargesharing, pixelerror, field anistropy."

I think you should double check with him to really confirm you are doing what you want.

**Comment by adrianpope** [comment]
I am confused about strategy:

- Originally I thought we were going to re-run the deep flats with the same effects enabled as last time, but with "backgamma 1" and see if that resulted in realistic B/F pixel correlation studies.

- But then it sounded like instead we first want to run in some mode to isolate pixel correlations from the background photon process, which would mean turning off sensor effects which can also correlate pixels. Here's where I got lost, because I have no idea how many different sensor effects there are, how they interact, and which ones might be critical to keep the tests from becoming too trivial.

I looked at [LSSTDESC/SSim_DC1#27](https://github.com/LSSTDESC/SSim_DC1/issues/27), especially [this comment](https://github.com/LSSTDESC/SSim_DC1/issues/27#issuecomment-276076875). It seems like guidance from @johnrpeterson is to use "fringing 0" and "cleardefects" to turn off sensor complications. So I looked through [source/raytrace/settings.cpp](https://bitbucket.org/phosim/phosim_release/src/cd0b8619b0f9013e3602edea6b60ad0512efef7b/source/raytrace/settings.cpp?at=v3.7.16&fileviewer=file-view-default#settings.cpp-595) to find out what "cleardefects" does in v3.7.16:
```
        if (keyName == "cleardefects") {
            impurityvariation = 0;
            fieldanisotropy = 0;
            deadlayer = 0;
            chargesharing = 0;
            pixelerror = 0;
        }
```
So that does cover everything that Chris listed, along with a few other things. From [this page about PhoSim v3.3](http://www.lsst-desc.org/FeaturedProjects/Phosim-v3.3) I suspect that impurityvariation might control tree rings, but the [PhoSim physics override commands documentation](https://bitbucket.org/phosim/phosim_release/wiki/Physics%20Commands) is a little terse for me to be confident about that.

It also sounds John recommends against "chargediffusion 0" because there would be "no sensor". I can find chargediffusion in the raytrace code, but I didn't see it in the physics override command doc, though "detectormode" is listed as having some affect on charge diffusion which I did not trace through in the code.

My best current guess for running without sensor complications is:
- Set "cleardefects" and "fringing 0", unless someone has a reason that we should be more fine-grained about the things included in cleardefects.
- Leave "chargediffusion" and "detectormode" at their defaults.
- Leave other commands that look possibly relate but are not included in "cleardefects" (eg. "saturation", "blooming") at their defaults, unless someone has a reason to change those.

I've also been looking at [DC2 Run 1.2p PhoSim Specifications](https://docs.google.com/document/d/1cn87h-LY2lyIxaUqWHQVFh2jGvbIf65Hz5hBwQUsqnU/edit) and I'm interested in understanding exactly what physics commands correspond to the tests described in the text. If someone has a pointer I'd appreciate it.

I'd like to confirm what we want to do in what order, and maybe I'll try to bring this up at the DM-DC2-TF or CI meetings tomorrow. Depending on settings, 6 pairs of flats takes ~50k-100k NERSC hours on Cori-KNL, so I'd like to minimize throw-away sets.

**Comment by johnrpeterson** [comment]
yes, all those settings are correct.  however, Pierre already studied BF in phosim flats with gamma=1.  (and gamma=1 was where we would have figured out how to do the BF in the first place, so there is only a question about what happens when gamma is not equal to 1).  was there any specific result or something you need on this (since its a lot of CPU hours)?

**Comment by TomGlanzman** [comment]
@adrianpope Please refer to github issue #140 for additional detail about the background tests.  Let me know if you have further questions.

**Comment by adrianpope** [comment]
@TomGlanzman : Thanks for the link. The info for tests 25-27 is what I was looking for, eg. chargesharing was associated with B/F and fieldanisotropy was associated with TreeRings for those tests, and they were both toggled together.

@johnrpeterson : I think the criteria for enabling B/F for 2.0p/2.1p is that the PhoSim B/F curve resembles the HSC and ImSim curves for tests as in [this notebook](https://github.com/LSSTDESC/DC2-production/blob/u/wmwv/bf_kernel/Notebooks/make_brighter_fatter_kernel.ipynb) with the physics commands we would use for 2.0p/2.1p (current proposal is 1.2p + "backgamma 1"). If we cannot produce a suitable PhoSim B/F curve then we would presumably "turn off" B/F for 2.0p/2.1p.

I'm still not clear about the intermediate tests, so I plan to ask these questions at today's CI telecon:
- Exactly what physics commands do we want to use for intermediate tests?
- Will we analyze the outputs of the intermediate tests with @wmwv 's notebook, or some other pixel correlation method? Do we know what signal we expect for which physics commands?
- Do the intermediate tests need to be deep (eg. 70k-90k e-/pixel) in both filters (u, i), or can we start with one filter and possibly shallower (eg. ~10k-30k e-/pixel) flat pairs if things like B/F are "off"?

**Comment by johnrpeterson** [comment]
the kernel described in @wmwv 's notebook would be related to the auto-correlation which we've computed before.  ask Pierre for the most recent version.  he might have something he can translate into the same form.

you mainly just need "chargesharing 1" which does the BF effect.  in terms of depth, you will want to do exactly the same as the tests you are comparing to.  i think they grow linearly or quadratic-ly with photon count rate, so you'll want to do a similar brightness.

**Comment by katrinheitmann** [comment]
@TomGlanzman Can we close this issue for now? I think we we want to come back to B/F and phosim we would need to start a fresh discussion. Thanks!

**Comment by TomGlanzman** [comment]
Issue closed for now!
